File: client/src/main/java/org/evosuite/kex/KexClassLoader.java
package org.evosuite.kex;

import java.net.URL;
import java.net.URLClassLoader;

public class KexClassLoader extends URLClassLoader {
    private final ClassLoader loader;

    public KexClassLoader(URL[] urls) {
        super(urls, null);

        this.loader = KexClassLoader.class.getClassLoader();
    }

    @Override
    protected Class<?> findClass(String s) throws ClassNotFoundException {
        try {
            return super.findClass(s);
        } catch (ClassNotFoundException e) {
            return loader.loadClass(s);
        }
    }

}

File: client/src/main/kotlin/org/evosuite/kex/KexClassLoader.kt
package org.evosuite.kex

import org.evosuite.runtime.instrumentation.RuntimeInstrumentation
import java.net.URL
import java.net.URLClassLoader

class KexClassLoader(urls: Array<URL>) : URLClassLoader(urls, null) {
    private val loader = KexClassLoader::class.java.classLoader

    override fun loadClass(s: String): Class<*> {
        if (!RuntimeInstrumentation.checkIfCanInstrument(s)) {
            return loader.loadClass(s)
        }
        return super.loadClass(s)
    }

    override fun findClass(s: String): Class<*> =
        try {
            super.findClass(s)
        } catch (e: ClassNotFoundException) {
            loader.loadClass(s)
        }
}